src/main/java/fi/pohina/vinkkilista/App.java

```diff
@@ -106,27 +88,28 @@ public void startServer(int portNumber) {
                 String title = req.queryParams("title");
                 String url = req.queryParams("url");
                 String author = req.queryParams("author");
-                User creator = users.findById(getUserIdFromSession(req.session()));
-                Set<String> tags = tagParser.parse(req.queryParams("tags"));
+                User creator = requestUserManager.getSignedInUser(req);
+                Set<String> tagNames = tagParser.csvToSet(req.queryParams("tags"));
+                tagNames.add(tagService.tagFromUrl(url));
                 
-                if (Strings.isNullOrEmpty(title) || Strings.isNullOrEmpty(url)) {
-                    res.redirect("new");
-                    return "Bookmark could not be created";
-                } else {
-                    bookmarks.createBookmark(
-                            title,
-                            url,
-                            author,
-                            creator,
-                            tags);
+                Set<Tag> tags = tagService.findOrCreateTags(tagNames);
+                Bookmark created = bookmarkService.createBookmark(title, url, author, creator, tags);
+
+                if (created != null) {
```

Comment:
Perhaps `bookmarkService.createBookmark()` should just return `Boolean` as the returned `Bookmark` is not otherwise used? This could also be embedded in the logic, see above.

Reply:
I don't think returning the bookmark is a problem. If we would continue with this project it would be needed at some point anyways.